src/database-postgresql.cpp

```diff
@@ -0,0 +1,505 @@
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "delete_block",
+			"DELETE FROM blocks WHERE pos = $1::bigint", 1,
+			(const Oid *)readFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "read_pos", "SELECT pos FROM blocks", 0, NULL);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	PQclear(m_res);
+
+	infostream << "Postgres: SQL statements prepared." << std::endl;
+}
+
+void Database_PostgreSQL::dbconnect(void)
+{
+	DSTACK(__FUNCTION_NAME);
+
+	m_conn = PQconnectdb(m_conninfo.c_str());
+
+	if (PQstatus(m_conn) != CONNECTION_OK) 
+	       	throw BaseException(PQerrorMessage(m_conn));
+
+	infostream << "Postgres: Database connection made." << std::endl;
+}
+
+void Database_PostgreSQL::connectString(void)
+{
+	DSTACK(__FUNCTION_NAME);
+
+	if (m_conninfo.length() > 1)
+		return;
+
+	try {
+		// username
+		if (m_conf.exists("pg_username")) {
+		    m_user = m_conf.get("pg_username");
+		} else {
+		    m_user = "minetest";
+		    infostream << "Postgres: Using default user, minetest.\n";
+		}
+
+		// Postgres host
+		if (m_conf.exists("pg_host"))
+		    m_host = m_conf.get("host");
+
+		// Port number
+		if (m_conf.exists("pg_port") && m_conf.exists("pg_host")) {
+		    m_port = m_conf.get("pg_port");
+		} else if (m_conf.exists("pg_port") && !m_conf.exists("pg_host")) {
+		    m_port = m_conf.get("pg_port");
+		    m_host = std::string("localhost");
+		} else if (!m_conf.exists("pg_port") && m_conf.exists("pg_host")) {
+		    m_port = "5432";
+		} else {
+		    infostream << "Postgres: Using socket connection.\n";
+		}
+
+		// password
+		if (m_conf.exists("pg_password"))
+		    m_password = m_conf.get("pg_password");
+		else
+		    infostream << "Using no password.\n";
+
+		// database name
+		if (m_conf.exists("pg_dbname")) {
+		    m_dbname = m_conf.get("pg_dbname");
+		} else {
+		    m_dbname = std::string("minetest");
+		    infostream << "Postges: Using default database name, minetest.\n";
+		}
+
+	} catch(SettingNotFoundException e) {
+		throw SettingNotFoundException("Set database in world.mt to use the pgsql backend");
+	}
+
+	m_conninfo = "user=" + m_user + " dbname=" + m_dbname;
+
+	if (m_host.size() > 0)
+		m_conninfo = m_conninfo + " host=" + m_host;
+
+	if (m_port.size() > 0)
+		m_conninfo = m_conninfo + " port=" + m_port;
+
+	if (m_password.size() > 0)
+		m_conninfo = m_conninfo + " password=" + m_password;
+
+	infostream << "Postgres connection info: " << m_conninfo << std::endl;
+}
+
+bool Database_PostgreSQL::initialized(void)
+{
+	return PQstatus(m_conn) == CONNECTION_OK ? true : false;
+}
+
+void Database_PostgreSQL::beginSave()
+{
+	DSTACK(__FUNCTION_NAME);
+
+	verifyDatabase();
+
+	m_res = PQexec(m_conn, "BEGIN;");
+
+	switch(PQresultStatus(m_res)) {
```

Comment:
if (PQresultStatus(m_res) != PGRES_COMMAND_OK) {
                throw BaseException(PQresultErrorMessage(m_res));
        }
rather than switch? Would that be okay?